staging/kos/pkg/controllers/connectionagent/connectionagent.go

```diff
@@ -0,0 +1,1059 @@
+		localityStr = "remote"
+	}
+	requeues := ca.queue.NumRequeues(attRef)
+	if err != nil {
+		glog.Warningf("Failed processing %s NetworkAttachment %s in VNI %d, requeuing (%d earlier requeues): %s",
+			localityStr,
+			attRef.nsn,
+			attRef.vni,
+			requeues,
+			err.Error())
+		ca.queue.AddRateLimited(attRef)
+		return
+	}
+	glog.V(4).Infof("Finished %s NetworkAttachment %s in VNI %d with %d requeues", localityStr, attRef.nsn, attRef.vni, requeues)
+	ca.queue.Forget(attRef)
+}
+
+func (ca *ConnectionAgent) processLocalAtt(nsn k8stypes.NamespacedName, vni uint32) error {
+	att, err := ca.localAttsLister.NetworkAttachments(nsn.Namespace).Get(nsn.Name)
+
+	if att != nil && att.Spec.VNI == vni {
+		// If we're here the attachment was found and the VNIs match, so we proceed
+		// to attempting to create a local interface.
+		firstLocalAttInVN := ca.updateVNStateForExistingLocalAtt(att)
+		if firstLocalAttInVN {
+			glog.V(2).Infof("VN with ID %d became relevant: an Informer has been started", att.Spec.VNI)
+		}
+		ifcNameForStatus, statusIfcNameMaybeNeedsUpdate, err := ca.attemptToCreateLocalAttIfc(att)
+		updatedAtt, statusUpdateErr := ca.setIfcNameAndHostIPInLocalAttStatusIfNeeded(att, ifcNameForStatus, statusIfcNameMaybeNeedsUpdate)
+		if updatedAtt != nil {
+			glog.V(2).Infof("Updated local att %s status with ifc name %s and host IP %s", nsn,
+				updatedAtt.Status.IfcName,
+				updatedAtt.Status.HostIP,
+			)
+		}
+		return aggregateErrors("\n\t", err, statusUpdateErr)
+	}
+
+	if (err != nil && k8serrors.IsNotFound(err)) || (att != nil && att.Spec.VNI != vni) {
+		// If we're here either the attachment has been deleted or its VNI is different wrt to the one
+		// in the reference. In both cases: (1) The vnState for the reference vni should be updated to
+		// reflect the fact that the local attachment is no longer there and (2) If an old interface (with
+		// the same VNI as the one in the reference) exists it should be deleted.
+		lastLocalAttInVN := ca.updateVNStateAfterLocalAttDeletion(nsn.Name, vni)
+		if lastLocalAttInVN {
+			glog.V(2).Infof("VN with ID %d became irrelevant: %s",
+				vni,
+				"its Informer has been stopped and references to remote atts in that VN have been enqueued")
+		}
+		deletedIfcName, ifcDeletionErr := ca.attemptToDeleteLocalAttIfc(nsn, vni)
+		if err != nil && k8serrors.IsNotFound(err) && ifcDeletionErr == nil {
+			// If we're here the attachment has been deleted and if its interface (associated with the reference vni)
+			// has been deleted as well. We do a best effort attempt to clear host IP and ifc name from deleted attachment status,
+			// no retries because the update might fail for a lot of legitimate reasons (e.g. the attachment has been deleted from
+			// etcd, the status has already been updated to a new value by the CA on the new node of the attachment, etc...). Ideally we
+			// would retry depending on the returned error, but it's a overhead and clearing the status is not that important.
```

Comment:
Why do exceptional error handling here?  The logic would be simpler with regular error handling.